man/man1/news.1
.TH NEWS 1
.SH NAME
news \- print news items
.SH SYNOPSIS
.B news
[
.B -a
]
[
.B -n
]
[
.I item ...
]
.SH DESCRIPTION
When invoked without options,
this simple local news service
prints files that have appeared in
.BR /lib/news
since last reading, most recent first,
with each preceded by an appropriate header.
The time of reading is recorded.
The options are
.TP
.B -a
Print all items, regardless of currency.
The recorded time is not changed.
.TP
.B -n
Report the names of the current items without
printing their contents, and without changing
the recorded time.
.PP
Other arguments
select particular news items.
.PP
To post a news item, create a file in
.BR  /lib/news .
.PP
You may arrange to receive news automatically by 
registering your mail address in
.BR /sys/lib/subscribers .
A daemon mails recent news
to all addresses on the list.
.PP
Empty news items, and news items named
.B core
or
.B dead.letter
are ignored.
.SH FILES
.TF /sys/lib/subscribers
.TP
.B /lib/news/*
articles
.TP
.B $HOME/lib/newstime
modify time is time news was last read
.TP
.B /sys/lib/subscribers
who gets news mailed to them
.SH SOURCE
.B /usr/local/plan9/src/cmd/news.c


CHANGES
November 26, 2005
	9term: add button 3 menu
	include: move utf.h, fmt.h from u.h to libc.h
	dist: print summary w/ errors during INSTALl

November 25, 2005
	root: add Makefile and configure for the blind,
		rename NOTES to README	

November 21, 2005
	lib9, 9pserve, vac: 9P2000.u fixes (Lucho Ionkov)

November 15, 2005
	libdraw: more X11 fixes
	libc.h: sched_yield NetBSD tweak (Anselm Garbe)

November 12, 2005
	faces: fix face location
	jpg, strings: remove references to /fd/

November 12, 2005
	various: X11 fixes

November 11, 2005
	9term: more fixes (Tim Wiess)

November 10, 2005
	9term: handle \r, other fixes (Tim Wiess)
	auxstats: new Darwin tweaks

November 9, 2005
	libthread: fix OS X context switch code

November 7, 2005
	vbackup: various fixes

November 6, 2005
	9term: new port directly from rio
	include/libc.h: p9syslog, not syslog in varargck
	libauthsrv: close fd, memory leak in authdial
	libdraw: X11 debugging helps
	lib9: add getcallerpc-arm (Ron Minnich)
	auxstats: Fix FreeBSD 6 build (William Josephson)

November 4, 2005
	stats: add -C to argchars (Lou Kamenov)
	libthread: ARM fixes (Ron Minnich), getcontext fixes

November 2, 2005
	vbackup: longer timeout on BSD mount
	libthread: Linux ARM fixes (Ron Minnich)

November 1, 2005
	various: ctype sign fixes (Tom Miller)
	libfmt: avoid overflow on %f
	libthread: Linux ARM fixes (Ron Minnich), FreeBSD 5 fixes
	libventi: various cleanup

October 31, 2005
	acme: preparation for multiline tags (Paul Lalonde)
	bc: 64-bit safe (Taj Khattra)
	faces: initial port (John Cummings)
	stats: add -F option (Lou Kamenov)
	various: ctype sign fixes (Tom Miller)

October 30, 2005
	auxstats: more OS X fixes (Jeff Sickel)	

October 29, 2005
	9p: clunk fids before exiting
	mk: handle MKSHELL type correctly
	upas, acme mail: add initial port (John Cummings)
	vbackup: silence compiler warnings
	venti: silence compiler warnings
	lib9: remove reference to ffork
		print correct file name in truerand error
	libdisk: add Disk routines
	libdiskfs: silence compiler warnings
	libframe: add noredraw flag
	libhtml: many fixes
	libsunrpc: silence compiler warnings

October 18, 2005
	9l: add autoframework (like autolib) for OS X (Jeff Sickel)
	page: use "9 file" always
	ps, psu: narrower display
	stats: better OS X support (Jeff Sickel)

October 17, 2005
	cb: undef old type macros before redefining.

October 5, 2005
	stats: better OS X support (Jeff Sickel) 

September 30, 2005
	libhtml: more &names and numbers (Erik Quanstrom)

September 29, 2005
	libc.h: add #pragmas for varargck, #ifdef'ed out.

September 26, 2005
	libthread: print thread info on SIGQUIT, SIGINFO, logging
	acme: add more thread names

September 21, 2005
	9term: fix SunOS typo (Bengt Kleberg, glaive)

September 19, 2005
	libhtml: more &names (Erik Quanstrom)

September 19, 2005
	lib9/truerand: OpenBSD lets you open /dev/random
		but not read it! (Chad Dougherty, Mechiel Lukkien)

September 15, 2005
	libdraw/x11-event: Ignore unexpected events (Rob Pike)
	jpg: do not revert to CMAP for second file (Rob Pike)

September 13, 2005
	9pserve: signedness bug

September 12, 2005
	dist: various attempts at Debian packaging
	lib/moveplan9.sh: allow specification of old directory
	Getdir: interpret tags with $ in them
	mpm: use g++ as LD, add tmac.pm to tmac directory
	include/libc.h, fcall.h: add 9P2000.u extensions
	9pserve: add 9P2000.u extensions
	vac: add 9P2000.u extensions, signedness bug

September 9, 2005
	lc: add -p like man page says
	various: insert (uchar) casts for ctype arguments
		for broken NetBSD.

September 4, 2005
	dist/main.html: note x86 OpenBSD

August 31, 2005
	file: spell IA-64 correctly (Zoltan Jarai)

August 30, 2005
	dist/debian, dist/pkg: start playing with building packages
	dist/isum: ignore 'loop not entered at top' (SunOS cc)
	dist/ss.html: add screen shots
	dist/unix.html: update VN license summary
	9l: no need to handle OpenBSD specially
	file: handle #! scripts, ELF better
	font/lucidasans/bold*unicode.7.font: more subfonts
	g: search .pl, .cgi
	hget: add hget(1)
	lp: add preliminary lp (not expected to work yet)
	src: use Plan 9 file(1)

August 12, 2005
	lib/moveplan9.files: add bin/stack
	include/libc.h: use DM constants from v9fs, add DMTMP
	libthread: add Linux-power-asm.s
	cb: add

August 11, 2005
	various: silence warnings
	unix/*: note new BSD license from Vita Nuova
	libdiskfs: add UFS2 support
	libthread: add NetBSD power assembly
	9term: send tty's current interrupt character on DEL keystroke
	rc: add limited rfork command, make sure errors propagate
	venti(8): document bloom line in config
	vftp: add get command
	vnfs: use strtoull for arguments instead of atoi.

August 10, 2005
	src/cmd/mkfile: do not build lp yet
	dist/addsrclinks, fixurls: use /usr/bin/env perl to find perl (NetBSD)
	lib9/_p9dir: avoid 32-bit overflow in old Linux size calculation
	lib9/truerand: use /dev/srandom on OpenBSD
	libthread: annoying wrappers for OpenBSD
	libmp: different assembly comments for old GNU as
	9p: set exit status
	file: add, appease gcc
	getflags: added
	mk: decide mtime-related race conservatively
	postscript/text2post: added
	rcmain: make fn cd safer for use with rc -e (in mk)
	rio: turn off debugging prints, remove 9wm.man, fixes from Andrey
	vbackup: NetBSD fixes for mount-BSD.c
	venti/srv: add <sys/stat.h> header for FreeBSD

July 28, 2005
	acme: subtle winunlock bug fix
	libdiskfs: fix double, triple indirect block reading for FFS
	vftp: new program
	vbackup: coalesce mount-*BSD files.

July 27, 2005
	include/u.h, src/libthread/thread.c: SunOS 10 attempt
	libdraw: less graceful, less raceful shutdown
	libthread: avoid C library races during threadexitsall
	awd: cannot use hostname -s on SunOS
	9pserve, srv: use correct type for NOFID

July 26, 2005
	various: fixes for SunOS from Bengt Kleberg

July 25, 2005
	vac: avoid infinite redirection while sniffing type of archive

July 24, 2005
	venti: many minor changes, updates

src/cmd/news.c
/*
 *	news foo	prints /lib/news/foo
 *	news -a		prints all news items, latest first
 *	news -n		lists names of new items
 *	news		prints items changed since last news
 */

#include <u.h>
#include <libc.h>
#include <bio.h>

#define	NINC	50	/* Multiples of directory allocation */
char	*NEWS = "#9/news";
char	TFILE[] = "%s/lib/newstime";

/*
 *	The following items should not be printed.
 */
char*	ignore[] =
{
	"core",
	"dead.letter",
	0
};

typedef
struct
{
	long	time;
	char	*name;
	vlong	length;
} File;
File*	n_list;
int	n_count;
int	n_items;
Biobuf	bout;

int	fcmp(const void *a, const void *b);
void	read_dir(int update);
void	print_item(char *f);
void	eachitem(void (*emit)(char*), int all, int update);
void	note(char *s);

void
main(int argc, char *argv[])
{
	int i;

	NEWS = unsharp(NEWS);

	Binit(&bout, 1, OWRITE);
	if(argc == 1) {
		eachitem(print_item, 0, 1);
		exits(0);
	}
	ARGBEGIN{
	case 'a':	/* print all */
		eachitem(print_item, 1, 0);
		break;

	case 'n':	/* names only */
		eachitem(note, 0, 0);
		if(n_items)
			Bputc(&bout, '\n');
		break;

	default:
		fprint(2, "news: bad option %c\n", ARGC());
		exits("usage");
	}ARGEND
	for(i=0; i<argc; i++)
		print_item(argv[i]);
	exits(0);
}

int
fcmp(const void *a, const void *b)
{
	long x;

	x = ((File*)b)->time - ((File*)a)->time;
	if(x < 0)
		return -1;
	if(x > 0)
		return 1;
	return 0;
}

/*
 *	read_dir: get the file names and modification dates for the
 *	files in /usr/news into n_list; sort them in reverse by
 *	modification date.
 */
void
read_dir(int update)
{
	Dir *d;
	char newstime[100], *home;
	int i, j, n, na, fd;

	n_count = 0;
	n_list = malloc(NINC*sizeof(File));
	na = NINC;
	home = getenv("HOME");
	if(home) {
		sprint(newstime, TFILE, home);
		d = dirstat(newstime);
		if(d != nil) {
			n_list[n_count].name = strdup("");
			n_list[n_count].time =d->mtime-1;
			n_list[n_count].length = 0;
			n_count++;
			free(d);
			if(n_list[n_count-1].name == nil){
				fprint(2, "Memory allocation failed");
				exits("Memory allocation failed");
			}
		}
		if(update) {
			fd = create(newstime, OWRITE, 0644);
			if(fd >= 0)
				close(fd);
		}
	}
	fd = open(NEWS, OREAD);
	if(fd < 0) {
		fprint(2, "news: ");
		perror(NEWS);
		exits(NEWS);
	}

	n = dirreadall(fd, &d);
	for(i=0; i<n; i++) {
		for(j=0; ignore[j]; j++)
			if(strcmp(ignore[j], d[i].name) == 0)
				goto ign;
		if(na <= n_count) {
			na += NINC;
			n_list = realloc(n_list, na*sizeof(File));
		}
		n_list[n_count].name = strdup(d[i].name);
		n_list[n_count].time = d[i].mtime;
		n_list[n_count].length = d[i].length;
		n_count++;
	ign:;
	}
	free(d);

	close(fd);
	qsort(n_list, n_count, sizeof(File), fcmp);
}

void
print_item(char *file)
{
	char name[4096], *p, *ep;
	Dir *dbuf;
	int f, c;
	int bol, bop;

	sprint(name, "%s/%s", NEWS, file);
	f = open(name, OREAD);
	if(f < 0) {
		fprint(2, "news: ");
		perror(name);
		return;
	}
	strcpy(name, "...");
	dbuf = dirfstat(f);
	if(dbuf == nil)
		return;
	Bprint(&bout, "\n%s (%s) %s\n", file,
		dbuf->muid[0]? dbuf->muid : dbuf->uid,
		asctime(localtime(dbuf->mtime)));
	free(dbuf);

	bol = 1;	/* beginning of line ...\n */
	bop = 1;	/* beginning of page ...\n\n */
	for(;;) {
		c = read(f, name, sizeof(name));
		if(c <= 0)
			break;
		p = name;
		ep = p+c;
		while(p < ep) {
			c = *p++;
			if(c == '\n') {
				if(!bop) {
					Bputc(&bout, c);
					if(bol)
						bop = 1;
					bol = 1;
				}
				continue;
			}
			if(bol) {
				Bputc(&bout, '\t');
				bol = 0;
				bop = 0;
			}
			Bputc(&bout, c);
		}
	}
	if(!bol)
		Bputc(&bout, '\n');
	close(f);
}

void
eachitem(void (*emit)(char*), int all, int update)
{
	int i;

	read_dir(update);
	for(i=0; i<n_count; i++) {
		if(n_list[i].name[0] == 0) {	/* newstime */
			if(all)
				continue;
			break;
		}
		if(n_list[i].length == 0)		/* in progress */
			continue;
		(*emit)(n_list[i].name);
	}
}

void
note(char *file)
{

	if(!n_items)
		Bprint(&bout, "news:");
	Bprint(&bout, " %s", file);
	n_items++;
}
